### DISCUSSION TITLE
More tightly scoped personal access tokens

### DISCUSSION THREAD

**Discussion by simonw** [community/community#10937] [Other Feature Feedback, Questions, & Ideas]
Personal access tokens are fantastic - they're such a low-friction way to build automation against my GitHub account. I use them for all sorts of things.

There's just one catch: the permissions I can assign to those tokens are pretty broad, at https://github.com/settings/tokens/new - I need to provide "repo" access with "Full control of private repositories" for a token that I only want to be able to write to a single one of my public repositories!

I'd really appreciate a lot more power here. I'd like to be able to issue tokens that can only interact with a specific subset of my repositories - I'd also like to have much more finely grained permissions control over what they can do ("full control of private repos" is way more than I want to grant something that I just need to able to make the occasional write request to the repository API).

I understand this is a HUGE undertaking - the permissions system associated with personal access tokens hasn't had any major changes in all of the time I've been using them, so I imagine this would take an extraordinary amount of engineering work. But I would appreciate it so much!

**Comment by simonw** [comment]
https://blog.thea.codes/building-a-stateless-api-proxy/ is a really fun workaround for this issue - Thea Flowers built a API proxy for GitHub that allowed her to issue her own encrypted tokens which could be scoped down to only allow access to specific endpoints. I'd rather not have to run that myself though!

**Comment by ckuethe** [comment]
So. Much. This.

I'm a member of multiple organizations. I'd like to make a number access tokens, which I could then restrict to the repositories belonging to some set of organizations (eg. only public repos that an anonymous user could see; organization access only - FooCorp, FooCorp+BazLLC, ...; or organization and public repos).

Then, when I'm using FooCorp token, to access FooCorp repositories, from FooCorp systems I could restrict that token to only operate in the FooCorp account, not even non-FooCorp public repos. Let's say FooCorp establishes some sort of business relationship with BazLLC, and doing some collaborative work with BazLLC is now in scope for me, I'd like to make a new token to use for that work.

I'd probably make a bunch of tokens for my own personal projects too.
